src/Ouroboros/Storage/Immutable/DB.hs

```diff
@@ -282,45 +296,192 @@ modifyInternalState db newEpoch action = do
+          hGet eHnd (fromEnum blobSize)
+
+    -- Steps the internal state of an 'Iterator'. Internal use only.
+    iteratorStep :: (MonadSTM m, MonadMask m, HasFSE m)
+                 => ExceptT ImmutableDBError m ()
+    iteratorStep = liftFsError $ do
+        ((epoch, RelativeSlot slot), fullIndex) <- atomically $ do
+            (,) <$> readTVar _it_current <*> readTVar _it_epoch_index
+
+        let epoch' = epoch + 1
+        -- If the current relative slot is equal to the size of the index, it
+        -- means we reached the end of the epoch and we need to open the next
+        -- epoch file.
+        case fromIntegral (slot + 1) == U.length fullIndex &&
+             not (outOfBounds (epoch', 0) _it_end) of
+             True  -> do
+                 eHnd <- atomically $ readTVar _it_epoch_handle
+                 eHnd' <- hOpen (_dbFolder db <> renderFile "epoch" epoch') ReadMode
+                          `finally` hClose eHnd
+                 index' <- loadIndex db epoch'
+                 atomically $ do
+                     writeTVar _it_epoch_handle eHnd'
+                     writeTVar _it_epoch_index index'
+                     writeTVar _it_current (epoch', RelativeSlot 0)
+             False -> atomically $ writeTVar _it_current (epoch, RelativeSlot $ slot + 1)
+
+-- | Dispose of the 'Iterator'.
+iteratorClose :: (MonadSTM m, MonadMask m, HasFSE m)
+              => Iterator m
+              -> m (Either ImmutableDBError ())
+iteratorClose Iterator{..} = runExceptT $ liftFsError $ do
+    atomically (readTVar _it_epoch_handle) >>= hClose
+
+-- | When given a start and a stop position, this function returns an
+-- 'Iterator' suitable to be used to efficiently stream binary blocks out of
+-- the 'ImmutableDB'.
+streamBinaryBlob :: forall m. (HasFSE m, MonadSTM m)
+                 => ImmutableDB m
+                 -> (Epoch, RelativeSlot)
+                 -- ^ When to start streaming.
+                 -> (Epoch, RelativeSlot)
+                 -- ^ When to stop streaming.
+                 -> m (Either ImmutableDBError (Iterator m))
+streamBinaryBlob db@ImmutableDB{..} start@(epoch,RelativeSlot slot) end = do
+    runExceptT $ liftFsError $ do
+        eHnd      <- hOpen (_dbFolder <> renderFile "epoch" epoch) ReadMode
+        fullIndex <- loadIndex db epoch
+
+        -- Position the epoch handle at the right place
+        _ <- hSeek eHnd AbsoluteSeek (fst $ fullIndex U.! fromIntegral slot)
+
+        (current, eVar, iVar) <- atomically $ do
+            (,,) <$> newTVar start <*> newTVar eHnd <*> newTVar fullIndex
+
+        return $ Iterator {
+            _it_current      = current
+          , _it_end          = end
+          , _it_epoch_handle = eVar
+          , _it_epoch_index  = iVar
+          }
+
+getBinaryBlob :: forall m. (HasFSE m, MonadSTM m, MonadMask m)
               => ImmutableDB m
               -> (Epoch, RelativeSlot)
               -> m (Either ImmutableDBError ByteString)
```

Comment:
Let's return `Maybe` here, since the requested slot might not have a block. 

Also, let's make it clear (comment) that although we are abstract over the blobs we are storing, we cannot store empty blobs.